### DISCUSSION TITLE
Send a deadline as well as a timeout

### DISCUSSION THREAD

**Discussion by fables-tales** [BattlesnakeOfficial/feedback#100] [Feature Requests]
# Brief Line, Up Front

Today the [`game`](https://docs.battlesnake.com/references/api#game) object sends a target timeout that a given game is expecting. This is a relative time but doesn't come with a basis, so some guesswork is required to work out how long a given request should process for. I think it would be cool for the battlesnake server to send a deadline, and a request origination time, to make it easier to work out when a given server needs to return

# Implementation Detail

Modify the game object to look like:
```json
{
  "id": "totally-unique-game-id",
  "ruleset": {
      "name": "standard",
      "version": "v1.2.3"
    },
  "timeout": 500 // to prevent breakage
  // new
  "timing_context": {
    // microsecond unix timestamp
    "sent_at_micros": "16297224781234000000" // note, unix timestamp as string microseconds, json doesn't have integers. at this resolution this will need to be a string to be communicated precisely.
    "deadline_micros": "16297224781234500000" // deadline is 500ms later
  }
}
```

## Why?

I think this would be a clarifying signal to snake devs that enables two things more cleanly than is possible today:

1. Latency between the battlesnake server and their own server: by doing the equivalent of `Instant::now() - sent_at_micros`, we get a clean estimation of the latency to the battlesnake server, enabling reasonable estimation of when to stop request processing
2. The actual time the battlesnake server is expecting the response back becomes totally clear. Right now there is some guesswork: presumably a timestamp is generated internally the server right as the request is sent, but it's also not clear how exactly it does that, this would clarify that greatly

## Alternatives

* Do nothing: reasonable, given that the timeout already provides a decent amount of information, if it would be high effort to provide this information, then  it's totally reasonable to close this discussion
* Send milis instead of micros: the amount of information in microseconds means that we can't reasonably transmit json integers, so using a milis may be easier
* Send nanos instead of micros: I picked micros as a tradeoff between resolution and complexity of representation. Nanos are slowly becoming the "standard" for distributed systems to communicate to each other, and so we could use this if we wanted.

Anyway, this is the start of the discussion, really would love thoughts back :D

**Comment by chris-bsnake** [comment]
I really like this concept.  Definitely a feature for more advanced Battlesnake devs to take advantage of, so I like how the simplified timeout parameter is left int the payload.

**Comment by robbles** [comment]
Is there a risk that clock drift between the server running the game and the snake dev's server throws off the calculation? Or is that something we'd just assume you deal with yourself however you see fit?

That might factor into the resolution a bit too - is there a point in using nanos vs. micros if the most accurate time remaining you could estimate from this might be closer to the micros range?
